packages/fxa-graphql-api/src/auth/gql-customs.guard.ts

```diff
@@ -1,35 +1,49 @@
 /* This Source Code Form is subject to the terms of the Mozilla Public
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
-import { CanActivate, ExecutionContext, Injectable } from '@nestjs/common';
+import {
+  CanActivate,
+  ExecutionContext,
+  Inject,
+  Injectable,
+  mixin,
+  Type,
+} from '@nestjs/common';
 import { GqlExecutionContext } from '@nestjs/graphql';
 import { Request } from 'express';
-import { CustomsService } from 'fxa-shared/nestjs/customs/customs.service';
-
+import { CustomsService } from '@fxa/shared/nestjs/customs';
 import { SessionTokenResult } from './session-token.strategy';
 
-@Injectable()
-export class GqlCustomsGuard implements CanActivate {
-  constructor(private customs: CustomsService) {
-    if (!customs) {
-      throw new Error('No customs service provided.');
+export function GqlCustomsGuard(action: string): Type<CanActivate> {
+  // This is a way to get injection and pass a parameter to the guard.
+  @Injectable()
+  class CustomsGuard implements CanActivate {
+    constructor(@Inject(CustomsService) private customs: CustomsService) {
+      if (!customs) {
+        throw new Error('No customs service provided.');
+      }
     }
-  }
 
-  async canActivate(context: ExecutionContext): Promise<boolean> {
-    const ctx = GqlExecutionContext.create(context);
-    const req = ctx.getContext().req as Request;
-    const { email } = (req.user as SessionTokenResult).session;
-    const handlerName = context.getHandler().name;
-    const className = (context as any).constructorRef.name;
-    const customsName = [className, handlerName].join('.');
-    await this.customs.check({
-      action: customsName,
-      email,
-      ip: req.ip ?? '', // req.ip may be undefined, but the original contract expects a string
-      headers: req.headers,
-      query: req.query,
-    });
-    return true;
+    async canActivate(context: ExecutionContext): Promise<boolean> {
+      const ctx = GqlExecutionContext.create(context);
+      const req = ctx.getContext().req as Request;
+      const { ip, headers, query } = req;
+      const { email, uid } = (req.user as SessionTokenResult)?.session || {
+        email: undefined,
+        uid: undefined,
+      };
```

Comment:
Interesting that the old code didn't need the optional chaining and the fallback value.  I guess `session` always exists according to the type?  Not sure if that's the case at runtime but I don't remember seeing errors related to that.

Reply:
I think it's always there, but at some point I'm pretty sure the linter or the tsc barked at me.

Reply:
Just tested this again... I'm full it. Cleaning this up.